clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -4781,6 +4782,7 @@ CodeGenModule::CreateRuntimeFunction(llvm::FunctionType *FTy, StringRef Name,
         }
       }
       setDSOLocal(F);
+      markRegisterParameterAttributes(F);
```

Comment:
This really shouldn't work this way: we should go through CodeGenModule::SetLLVMFunctionAttributes to get all the correct attributes for a function.  Unfortunately, that's a bit complicated in this context because you need to construct a CGFunctionInfo, and none of the callers currently do that.  Which might be more work than you really want...  CC @JonPsson1 @jdoerfert @jhuber6

Reply:
I was trying to basically duplicate what was already done for the libcall functions. What is gained by using CGFunctionInfo? (Things already work as-is -- though generally it does looks like -mregparm support is kinda ad-hoc.)

Reply:
Oh, I think I see what you mean -- this is the common place where all functions should get their attributes set. Can the libcall code use this? (As in, can we remove `markRegisterParameterAttributes()` entirely and move the logic into `SetLLVMFunctionAttributes()` or is that just fantastic overkill?

Reply:
Zero/sign-extend attributes are also missing, I think.  Which probably doesn't affect x86, but could have obscure effects on some targets.

Using SetLLVMFunctionAttributes here isn't really a problem, except that it takes a clang::Type, not an llvm::Type type, and we only have a conversion in the other direction.  So you'd need to modify the callers.  And there are a lot of callers.

Reply:
To clarify, the code in llvm/ still needs to exist because we need some approximation of the calling convention even if we don't have access to the clang AST.  But clang shouldn't use it; clang's native representation of calling conventions is better.

(It's a long-standing request to try to give LLVM a representation of calling conventions that's closer to clang's representation, but it's a large problem, and nobody has really made any progress on it.)

Reply:
This seems like a large proposed change; is it worth it for this corner case? I think I would prefer to fix this as I have it now.

Reply:
Sign/zero-extend attributes led to real issues on some targets in the past, so it's not unlikely someone will need to address it at some point.  But maybe it doesn't need to be here.  I guess leave a FIXME noting it as a known issue.

I'm not really happy with exposing this interface on BuildLibCalls.h, since it's really a pretty big footgun (due to the zero/sign-extension issue I mentioned).  But copy-pasting the code is worse, so... I guess leave a comment in the header.

Reply:
Comment added. Is this what you had in mind?

Reply:
Suggested text on the clang change:

FIXME: We should use CodeGenModule::SetLLVMFunctionAttributes() instead of trying to approximate the attributes using the LLVM function signature.  This requires revising the API of CreateRuntimeFunction().

Reply:
Ah-ha, thanks! Okay, I've updated the comments with just a minor tweak.